File: javascript/packages/linter/src/rules/erb-no-output-control-flow.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { ParserRule, BaseAutofixContext, Mutable } from "../types.js"

import type { ParseResult, ERBIfNode, ERBUnlessNode, ERBElseNode, ERBEndNode, ERBIterationBlockNode, ERBBlockNode, ParserOptions } from "@herb-tools/core"
import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig } from "../types.js"

type OutputControlFlowNode = ERBIfNode | ERBUnlessNode | ERBElseNode | ERBEndNode | ERBIterationBlockNode | ERBBlockNode

interface ERBNoOutputControlFlowAutofixContext extends BaseAutofixContext {
  node: Mutable<OutputControlFlowNode>
}

class ERBNoOutputControlFlowRuleVisitor extends BaseRuleVisitor<ERBNoOutputControlFlowAutofixContext> {
  visitERBIfNode(node: ERBIfNode): void {
    this.checkOutputControlFlow(node)
    this.visitChildNodes(node)
  }

  visitERBUnlessNode(node: ERBUnlessNode): void {
    this.checkOutputControlFlow(node)
    this.visitChildNodes(node)
  }

  visitERBElseNode(node: ERBElseNode): void {
    this.checkOutputControlFlow(node)
    this.visitChildNodes(node)
  }

  visitERBEndNode(node: ERBEndNode): void {
    this.checkOutputControlFlow(node)
    this.visitChildNodes(node)
  }

  visitERBIterationBlockNode(node: ERBIterationBlockNode): void {
    this.checkOutputIterationBlock(node)
    this.visitChildNodes(node)
  }

  private checkOutputIterationBlock(node: ERBIterationBlockNode): void {
    const openTag = node.tag_opening

    if (!openTag) return
    if (openTag.value !== "<%=") return

    const method = node.message?.value ?? "each"

    const content = node.content?.value ?? " "
    const tagClosing = node.tag_closing?.value ?? "%>"
    const suggestion = `<%${content}${tagClosing}`.replace(/\s*\n\s*/g, " ")

    this.addOffense(
      `Iteration blocks like \`${method}\` should not be used with output tags, they return the collection instead of the rendered output. Use \`${suggestion}\` instead.`,
      openTag.location,
      { node: node as Mutable<ERBIterationBlockNode>, nodeType: "AST_ERB_BLOCK_NODE" },
    )
  }

  private static readonly CONTROL_BLOCK_NAMES: Record<string, string> = {
    "AST_ERB_IF_NODE": "if",
    "AST_ERB_ELSE_NODE": "else",
    "AST_ERB_END_NODE": "end",
    "AST_ERB_UNLESS_NODE": "unless"
  }

  private checkOutputControlFlow(controlBlock: ERBIfNode | ERBUnlessNode | ERBElseNode | ERBEndNode): void {
    const openTag = controlBlock.tag_opening;
    if (!openTag) {
      return
    }

    if (openTag.value === "<%="){
      const content = controlBlock.content?.value ?? ""
      const collapsedContent = content.replace(/\s*\n\s*/g, " ")
      const keyword = collapsedContent.trim().split(/\s+/)[0] || ERBNoOutputControlFlowRuleVisitor.CONTROL_BLOCK_NAMES[controlBlock.type] || controlBlock.type

      const tagClosing = controlBlock.tag_closing?.value ?? "%>"
      const suggestion = collapsedContent ? `<%${collapsedContent}${tagClosing}` : `<% ${keyword} ... ${tagClosing}`

      this.addOffense(
        `Control flow statements like \`${keyword}\` should not be used with output tags. Use \`${suggestion}\` instead.`,
        openTag.location,
        { node: controlBlock as Mutable<OutputControlFlowNode> },
      )
    }
  }
}

export class ERBNoOutputControlFlowRule extends ParserRule<ERBNoOutputControlFlowAutofixContext> {
  static ruleName = "erb-no-output-control-flow"
  static introducedIn = this.version("0.4.0")
  static autocorrectable = true

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      iteration_nodes: true
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense<ERBNoOutputControlFlowAutofixContext>[] {
    const visitor = new ERBNoOutputControlFlowRuleVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }

  autofix(offense: LintOffense<ERBNoOutputControlFlowAutofixContext>, result: ParseResult, _context?: Partial<LintContext>): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node } = offense.autofixContext

    if (!node.tag_opening) return null
    if (node.tag_opening.value !== "<%=") return null

    node.tag_opening.value = "<%"

    return result
  }
}

File: rust/herb-linter/src/rules/erb_no_output_control_flow.rs
use crate::utils::source_slice::collapse_newline_runs;
use herb::nodes::{ERBElseNode, ERBEndNode, ERBIfNode, ERBUnlessNode};
use herb::{Token, Visitor};

rule_visitor!(NoOutputControlFlowVisitor);
define_parser_rule!(ERBNoOutputControlFlowRule, "erb-no-output-control-flow", Error, NoOutputControlFlowVisitor,
  parser_options: { iteration_nodes: true },
  introduced_in: "0.4.0"
);

impl NoOutputControlFlowVisitor {
  fn check_output_control_flow(&mut self, tag_opening: Option<&Token>, content: Option<&Token>, tag_closing: Option<&Token>, fallback: &str) {
    let open_tag = match tag_opening {
      Some(token) if token.value == "<%=" => token,
      _ => return,
    };

    let collapsed = content.map(|token| collapse_newline_runs(&token.value)).unwrap_or_default();

    let keyword = collapsed
      .trim()
      .split_whitespace()
      .next()
      .filter(|keyword| !keyword.is_empty())
      .map(|keyword| keyword.to_string())
      .unwrap_or_else(|| fallback.to_string());

    let closing = tag_closing.map(|token| token.value.as_str()).unwrap_or("%>");

    let suggestion = if collapsed.is_empty() {
      format!("<% {keyword} ... {closing}")
    } else {
      format!("<%{collapsed}{closing}")
    };

    self.add_offense(
      format!("Control flow statements like `{keyword}` should not be used with output tags. Use `{suggestion}` instead."),
      open_tag.location.clone(),
    );
  }

  fn check_output_iteration_block(&mut self, node: &herb::nodes::ERBIterationBlockNode) {
    let open_tag = match node.tag_opening.as_ref() {
      Some(token) if token.value == "<%=" => token,
      _ => return,
    };

    let method = node.message.as_ref().map(|token| token.value.as_str()).unwrap_or("each");
    let content = node.content.as_ref().map(|token| token.value.as_str()).unwrap_or(" ");
    let closing = node.tag_closing.as_ref().map(|token| token.value.as_str()).unwrap_or("%>");
    let suggestion = collapse_newline_runs(&format!("<%{content}{closing}"));

    self.add_offense(
      format!(
        "Iteration blocks like `{method}` should not be used with output tags, they return the collection instead of the rendered output. Use `{suggestion}` instead."
      ),
      open_tag.location.clone(),
    );
  }
}

impl Visitor for NoOutputControlFlowVisitor {
  fn visit_erb_iteration_block_node(&mut self, node: &herb::nodes::ERBIterationBlockNode) {
    self.check_output_iteration_block(node);
    self.walk_erb_iteration_block_node(node);
  }

  fn visit_erb_if_node(&mut self, node: &ERBIfNode) {
    self.check_output_control_flow(node.tag_opening.as_ref(), node.content.as_ref(), node.tag_closing.as_ref(), "if");
    self.walk_erb_if_node(node);
  }

  fn visit_erb_unless_node(&mut self, node: &ERBUnlessNode) {
    self.check_output_control_flow(node.tag_opening.as_ref(), node.content.as_ref(), node.tag_closing.as_ref(), "unless");
    self.walk_erb_unless_node(node);
  }

  fn visit_erb_else_node(&mut self, node: &ERBElseNode) {
    self.check_output_control_flow(node.tag_opening.as_ref(), node.content.as_ref(), node.tag_closing.as_ref(), "else");
    self.walk_erb_else_node(node);
  }

  fn visit_erb_end_node(&mut self, node: &ERBEndNode) {
    self.check_output_control_flow(node.tag_opening.as_ref(), node.content.as_ref(), node.tag_closing.as_ref(), "end");
    self.walk_erb_end_node(node);
  }
}
